File: csharp/protein-translation/ProteinTranslation.cs
using System;
using System.Text.RegularExpressions;
using System.Collections.Generic;

public static class ProteinTranslation
{
    private static string CodonSplit = "(?<=\\G.{3})";
    private static string StopValue = "STOP";
    private static string SkipValue = "SKIP";
    private static Dictionary<string, string> map = new Dictionary<string, string>()
    {
        { "AUG", "Methionine" },
        { "UUU", "Phenylalanine" },
        { "UUC", "Phenylalanine" },
        { "UUA", "Leucine" },
        { "UUG", "Leucine" },
        { "UCU", "Serine" },
        { "UCC", "Serine" },
        { "UCA", "Serine" },
        { "UCG", "Serine" },
        { "UAU", "Tyrosine" },
        { "UAC", "Tyrosine" },
        { "UGU", "Cysteine" },
        { "UGC", "Cysteine" },
        { "UGG", "Tryptophan" },
        { "UAA", "STOP" },
        { "UAG", "STOP" },
        { "UGA", "STOP" }
    };

    public static string[] Proteins(string strand)
    {
        string[] rnaParts = Regex.Split(strand, CodonSplit);
        var result = new List<string>();

        foreach(var entry in rnaParts)
        {
            string current = map.GetValueOrDefault(entry, SkipValue);

            if (current == StopValue)
            {
                break;
            }
            else if (current == SkipValue)
            {
                continue;
            }
            else
            {
                result.Add(current);
            }
        }

        return result.ToArray();
    }
}

File: javascript/protein-translation/protein-translation.js
const map = new Map();
map.set("AUG", "Methionine");
map.set("UUU", "Phenylalanine");
map.set("UUC", "Phenylalanine");
map.set("UUA", "Leucine");
map.set("UUG", "Leucine");
map.set("UCU", "Serine");
map.set("UCC", "Serine");
map.set("UCA", "Serine");
map.set("UCG", "Serine");
map.set("UAU", "Tyrosine");
map.set("UAC", "Tyrosine");
map.set("UGU", "Cysteine");
map.set("UGC", "Cysteine");
map.set("UGG", "Tryptophan");
map.set("UAA", "STOP");
map.set("UAG", "STOP");
map.set("UGA", "STOP");


export const translate = (rna) => {

    let rnaParts = rna.math(/[A-Z]{3}/g);
    let proteins = [];

    for (const part of rnaParts) {

      if (map.has(part)) {
        proteins.push(part);
      } else {
        throw new Error('Invalid codon');
      }
    }

    return proteins;
};
